src/utils/wrap_thread.hpp

```diff
@@ -0,0 +1,54 @@
+#ifndef WRAP_THREAD_HPP
+#define WRAP_THREAD_HPP
+
+#include <chrono>
+#include <future>
+#include <thread>
+
+/// More advanced then std::thread. Will join itself on destruction to avoid certain UBs. Has debug and stats.
+/// @owner hb
+class wrap_thread {
+	public:
+		wrap_thread() noexcept;
+
+		// std::chrono::duration destroy_time = 0
+		template<typename Function, typename ...Args>
+		explicit wrap_thread(Function&& f, Args&&... arg)
+			: wrap_thread()
+		{
+			m_time_started = t_clock::now();
+			m_future = std::async(std::launch::async, std::forward<Function>(f), std::forward<Args>(arg)...);
+
+		}
+		template<typename Function, typename ...Args>
+		explicit wrap_thread(std::chrono::seconds destroy_time, Function&& f, Args&&... arg)
+			: wrap_thread(std::forward<Function>(f), std::forward<Args>(arg)...) {
+
+			m_destroy_timeout=destroy_time;
+		}
+
+		wrap_thread(const wrap_thread &) = delete;
+
+		wrap_thread(wrap_thread && rhs) noexcept;
+
+		wrap_thread & operator=(wrap_thread && rhs) noexcept;
+
+		void join();
```

Comment:
we could also use:

bool try_join( chono::duration duration) ... 

it will wait at most around "duration" time, and if thread is not joined by then then it return false;

otherwise it returns true; if thread is joined (e.g. the thread finished, or it was an empty thread).